pytorch_lightning/metrics/functional/classification.py

```diff
@@ -872,6 +873,89 @@ def average_precision(
     return -torch.sum((recall[1:] - recall[:-1]) * precision[:-1])
 
 
+def object_detection_mean_average_precision(
+    pred: torch.Tensor, target: torch.Tensor, iou_threshold: float, num_classes: int
+) -> torch.Tensor:
+    """
+    Compute mean average precision for object detection task. Calculates average precision
+    using AUC.
+
+    Args:
+        pred: Tensor containing predictions, where each row is of the following format:
+              [image_idx, class_pred, class_prob, x_min, y_min, x_max, y_max]
+        target: Tensor containing ground truths, where each row is of the follwing format:
+                [image_idx, class_label, x_min, y_min, x_max, y_max]
+        iou_threshold: threshold for IoU score for determining true positive and
+                       false positive predictions.
+        num_classes: number of total classes
+
+    Returns:
+        mean of the average precision for each class in object detection task.
+    """
+    average_precisions = torch.zeros(num_classes)
+    for c in range(num_classes):
+        c_pred = [p for p in pred if p[1] == c]
+        c_target = [t for t in target if t[1] == c]
+        if len(c_target) == 0:
+            continue
+        # Sort c_pred in descending order w.r.t. confidence score
+        c_pred = sorted(c_pred, key=lambda x: x[2], reverse=True)
+        num_targets_per_image = Counter(t[0].item() for t in c_target)
+        targets_tracker = {
+            image_idx: torch.zeros(count)
+            for image_idx, count in num_targets_per_image.items()
+        }
+        tps, fps = torch.zeros(len(c_pred)), torch.zeros(len(c_pred))
+        for i, p in enumerate(c_pred):
+            # Get the targets that correspond to the same image as the prediction
+            ground_truths = [t for t in c_target if t[0] == p[0]]
+            best_iou = 0
+            best_target_idx = 0
+            for j, t in enumerate(ground_truths):
+                curr_iou = object_detection_iou(p[None, 3:], t[None, 2:])
+                if curr_iou > best_iou:
+                    best_iou = curr_iou
+                    best_target_idx = j
+            image_idx = int(p[0].item())
+            if (
+                best_iou > iou_threshold
+                and targets_tracker[image_idx][best_target_idx] == 0
+            ):
+                targets_tracker[image_idx][best_target_idx] = 1
+                tps[i] = 1
+            else:
+                fps[i] = 1
+        tps_cum, fps_cum = torch.cumsum(tps, dim=0), torch.cumsum(fps, dim=0)
+        precision = tps_cum / (tps_cum + fps_cum)
```

Comment:
This is actually a good point. 
It could happen that, for instance, you compute mAP before having seen the whole dataset and you haven't already seen ground truths for all classes. Or maybe you are using DDP and have the data split by class. 
In my implementation I solve this by doing so:
```python
class_ids = torch.cat((ground_truths[:, 6], detections[:, 6]), dim=0).unique()
```
Given that a bounding box is represented as ```[x1, y1, x2, y2, image_id, class_id, confidence]```

In general I think that the approach for each case, given the class_id, should be:
* No Detections, No Ground Truths -> this is just ignored as the class_id is not present in any way
* No Detections, Some Ground Truths -> AP = 0.0
* Some Detections, Some Ground Truths -> Normal Computation
* Some Detections, No Ground Truths -> Again Normal Computation where each Detection is a FP

Reply:
Great points. I added changes to handle all the scenarios that you've helpfully pointed out. I also added a test that verifies that it works correctly under such corner cases.